### PR TITLE
Implementation of bidirection kill for kill filter.

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message: Modify the kill filter to support directionality.
Additional Description: This allows us to more easily test response crashes.
Risk Level: low (this is a test-only filter AFAIK)
Testing: unit test, integration test
Docs Changes: Included
Release Notes: Included
Platform Specific Features: N/A
Related Issue #14249 

/assign @qqustc 

Since you're one of the maintainers of the filter. PTAL.

### CODE REVIEW

**Review by qqustc — APPROVED** [review]

**Comment by KBaichoo** [comment]
@alyssawilk since you're the maintainer on-call, can you assign maintainer to review these changes? @qqustc is an extension owner and has already signed off. Thanks!

File: source/extensions/filters/http/kill_request/kill_request_filter.cc [hunk 1]
```diff
@@ -15,25 +18,36 @@ namespace KillRequest {
...
       !absl::SimpleAtob(kill_request_header[0]->value().getStringView(), &is_kill_request)) {
+    return false;
+  }
+
+  return is_kill_request;
+}
+
+Http::FilterHeadersStatus KillRequestFilter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  bool is_correct_direction = kill_request_.direction() == KillRequest::INBOUND;
+  bool is_kill_request = isKillRequest(headers);
```
Comment (htuch) [inline source/extensions/filters/http/kill_request/kill_request_filter.cc:49]: Nit: `const bool`

File: source/extensions/filters/http/kill_request/kill_request_filter.cc [hunk 2]
```diff
@@ -46,11 +60,27 @@ Http::FilterHeadersStatus KillRequestFilter::decodeHeaders(Http::RequestHeaderMa
...
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus KillRequestFilter::encodeHeaders(Http::ResponseHeaderMap& headers, bool) {
+  if (kill_request_.direction() == KillRequest::INBOUND) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  bool is_kill_request = isKillRequest(headers);
```
Comment (htuch) [inline source/extensions/filters/http/kill_request/kill_request_filter.cc:82]: Ditto.

File: api/envoy/extensions/filters/http/kill_request/v3/kill_request.proto [hunk 3]
```diff
@@ -19,10 +19,19 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // Configuration for KillRequest filter.
 message KillRequest {
+  // On which direction should the filter check for the `kill_request_header`.
+  // Default to `INBOUND` if unspecified.
+  enum Direction {
+    INBOUND = 0;
+    OUTBOUND = 1;
```
Comment (htuch) [inline api/envoy/extensions/filters/http/kill_request/v3/kill_request.proto:26]: Question on whether to call this REQUEST/RESPONSE or DECODE/ENCODE?

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo minor comments.
/wait

File: api/envoy/extensions/filters/http/kill_request/v3/kill_request.proto (hunk 3 above)
Comment (antoniovicente) [reply api/envoy/extensions/filters/http/kill_request/v3/kill_request.proto:26]: decode and encode happen in each direction.  Vote for REQUEST/RESPONSE

Comment (KBaichoo) [reply api/envoy/extensions/filters/http/kill_request/v3/kill_request.proto]: Went with REQUEST/RESPONSE

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**